server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java

```diff
@@ -53,64 +113,198 @@
     @Override
     @ActionEvent(eventType = EventTypes.EVENT_ANNOTATION_CREATE, eventDescription = "creating an annotation on an entity")
     public AnnotationResponse addAnnotation(AddAnnotationCmd addAnnotationCmd) {
-        return addAnnotation(addAnnotationCmd.getAnnotation(), addAnnotationCmd.getEntityType(), addAnnotationCmd.getEntityUuid());
+        return addAnnotation(addAnnotationCmd.getAnnotation(), addAnnotationCmd.getEntityType(),
+                addAnnotationCmd.getEntityUuid(), addAnnotationCmd.isAdminsOnly());
     }
 
-    public AnnotationResponse addAnnotation(String text, EntityType type, String uuid) {
-        CallContext ctx = CallContext.current();
-        String userUuid = ctx.getCallingUserUuid();
+    public AnnotationResponse addAnnotation(String text, EntityType type, String uuid, boolean adminsOnly) {
+        UserVO userVO = getCallingUserFromContext();
+        String userUuid = userVO.getUuid();
+        checkAnnotationPermissions(userUuid, type, userVO);
 
-        AnnotationVO annotation = new AnnotationVO(text, type, uuid);
+        AnnotationVO annotation = new AnnotationVO(text, type, uuid, adminsOnly);
         annotation.setUserUuid(userUuid);
         annotation = annotationDao.persist(annotation);
         return createAnnotationResponse(annotation);
     }
 
+    private void checkAnnotationPermissions(String entityUuid, EntityType type, UserVO user) {
+        if (!isCallingUserAdmin()) {
```

Comment:
We could invert this first if and add a return (and maybe a log), to reduce code indentation.